backend/web/web_backend/src/cache.rs

```diff
@@ -1,28 +1,197 @@
+                return false;
+            }
+        }
+        false
+    }
 }
 
-impl<K: Eq + Hash + Clone, V: Clone> Cache<K, V> {
+// Emluate queue where we can save the Keys from cache
+// As long as the queue is not filled, we save all the data
+
+// When the queue hits maximum, we will start replacing the
+// oldest entries.
+impl MaxSize {
+    fn new(max_size: usize) -> Self {
+        MaxSize {
+            max_size,
+            filo: Vec::new(), // first-in-last-out
+            index: 0,
+        }
+    }
+
+    // If the cache is filled. Return a key to data that should be
+    // removed
+    fn put(&mut self, data: String) -> Option<String> {
+        // if the cache has not been filled yet
+        if self.filo.len() < self.max_size {
+            self.filo.push(data);
+            self.index += 1;
+            None
+        } else {
+            // If we are at the last index, set index to start position
+            if self.has_reached_max_size() {
+                self.index = 0;
+            }
+
+            let ret = replace(&mut self.filo[self.index], data);
+            self.index += 1;
+            Some(ret)
+        }
+    }
+
+    fn has_reached_max_size(&self) -> bool {
+        self.filo.len() == self.max_size
+    }
+}
+
+pub struct CacheBuilder {
+    max_size: Option<usize>,
+    timeout: Option<Duration>,
+}
+
+impl CacheBuilder {
     pub fn new() -> Self {
+        CacheBuilder {
+            timeout: None,
+            max_size: None,
+        }
+    }
+
+    pub fn with_max_size(self, max_size: usize) -> Self {
+        CacheBuilder {
+            max_size: Some(max_size),
+            timeout: self.timeout,
+        }
+    }
+
+    pub fn with_timeout(self, timeout: Duration) -> Self {
+        CacheBuilder {
+            max_size: self.max_size,
+            timeout: Some(timeout),
+        }
+    }
+
+    pub fn build<V>(self) -> Cache<V> {
         Cache {
             cache: RwLock::new(HashMap::new()),
+            timeout: self
+                .timeout
+                .map(|duration| RwLock::new(Timeout::new(duration))),
+            max_size: self.max_size.map(|size| RwLock::new(MaxSize::new(size))),
         }
     }
+}
+
+impl Default for CacheBuilder {
+    fn default() -> Self {
+        CacheBuilder::new()
+    }
+}
 
-    pub async fn lookup(&self, key: &K) -> Option<V> {
+pub struct Cache<V> {
+    cache: RwLock<HashMap<String, Arc<V>>>,
+    max_size: Option<RwLock<MaxSize>>,
+    timeout: Option<RwLock<Timeout>>,
+}
+
+impl<V: Clone> Cache<V> {
+    pub fn builder() -> CacheBuilder {
+        CacheBuilder::default()
+    }
+
+    pub async fn lookup(&self, key: &str) -> Option<Arc<V>> {
+        // If the timeout is set
+        if let Some(timeout) = &self.timeout {
+            // If the entry is too old
+            if !timeout.write().await.is_valid(String::from(key)) {
+                // TODO: if this is invalid, MaxSize will contain one unnecessary
+                // entry
```

Comment:
144 makes lookup grab a write lock for every lookup regardless if it is needed or not.
Is the TODO a problem? If it is not, remove the TODO, if it is, it should be resolved before the pull.